apollo-router/src/plugins/telemetry/config_new/events.rs

```diff
@@ -11,24 +22,348 @@ use crate::plugins::telemetry::config_new::extendable::Extendable;
+                }),
+            })
+            .collect();
+
+        SupergraphEvents {
+            request: self.supergraph.attributes.request,
+            response: self.supergraph.attributes.response,
+            error: self.supergraph.attributes.error,
+            custom: custom_events,
+        }
+    }
+
+    pub(crate) fn new_subgraph_events(&self) -> SubgraphEvents {
+        let custom_events = self
+            .subgraph
+            .custom
+            .iter()
+            .map(|(event_name, event_cfg)| CustomEvent {
+                inner: Mutex::new(CustomEventInner {
+                    name: event_name.clone(),
+                    level: event_cfg.level,
+                    event_on: event_cfg.on,
+                    message: event_cfg.message.clone(),
+                    selectors: event_cfg.attributes.clone().into(),
+                    condition: event_cfg.condition.clone(),
+                    attributes: Vec::new(),
+                }),
+            })
+            .collect();
+
+        SubgraphEvents {
+            request: self.subgraph.attributes.request,
+            response: self.subgraph.attributes.response,
+            error: self.subgraph.attributes.error,
+            custom: custom_events,
+        }
+    }
+}
+
+pub(crate) type RouterEvents =
+    CustomEvents<router::Request, router::Response, RouterAttributes, RouterSelector>;
+
+pub(crate) type SupergraphEvents = CustomEvents<
+    supergraph::Request,
+    supergraph::Response,
+    SupergraphAttributes,
+    SupergraphSelector,
+>;
+
+pub(crate) type SubgraphEvents =
+    CustomEvents<subgraph::Request, subgraph::Response, SubgraphAttributes, SubgraphSelector>;
+
+pub(crate) struct CustomEvents<Request, Response, Attributes, Sel>
+where
+    Attributes: Selectors<Request = Request, Response = Response> + Default,
+    Sel: Selector<Request = Request, Response = Response> + Debug,
+{
+    request: EventLevel,
+    response: EventLevel,
+    error: EventLevel,
+    custom: Vec<CustomEvent<Request, Response, Attributes, Sel>>,
+}
+
+impl Instrumented
+    for CustomEvents<router::Request, router::Response, RouterAttributes, RouterSelector>
+{
+    type Request = router::Request;
+    type Response = router::Response;
+
+    fn on_request(&self, request: &Self::Request) {
+        if self.request != EventLevel::Off {
+            let mut attrs = HashMap::with_capacity(5);
+            #[cfg(test)]
+            let mut headers: indexmap::IndexMap<String, HeaderValue> = request
+                .router_request
+                .headers()
+                .clone()
+                .into_iter()
+                .filter_map(|(name, val)| Some((name?.to_string(), val)))
+                .collect();
+            #[cfg(test)]
+            headers.sort_keys();
+            #[cfg(not(test))]
+            let headers = request.router_request.headers();
+
+            attrs.insert("http.request.headers".to_string(), format!("{:?}", headers));
```

Comment:
do we have to do that for every request if events are active, even if the configuration does not ask for the headers?

Reply:
no, only when you enabled `events.x.request|response` by default it's disabled